feat!: publish as esm-only

achingbrain opened the pull request:
Following on from the conversation in https://github.com/ipld/ipld/issues/224 this converts this repo to use the latest aegir with Unified CI.

1. Remove all build related dev deps apart from `aegir`
1. Run the `npx aegir check-project` command to update project config
1. Remove non-Unified-CI github actions
1. Moved source files into `src`
1. Rename `test/*.js` to `test/*.spec.js` so aegir can find them
1. Update `tsconfig.json` to extend config from `aegir`
1. Remove `"main"` and other unused fields from `package.json`
1. Use `chai` from `aegir` pre-configured with plugins we use
1. Fixes everything the linter complains about

This will need a follow up PR to `protocol/.github` to add this repo to the Unified CI [config file](https://github.com/protocol/.github/blob/master/configs/js.json) so it'll get automated config updates in the future.

Apologies that this PR is so noisy, most of it is from the `check-project` command

achingbrain commented:
Refs https://github.com/ipld/ipld/issues/224

rvagg commented on src/indexed-reader.js:
```diff
@@ -68,7 +68,7 @@ export class CarIndexedReader {
    * See {@link CarReader#getRoots}
    *
-   * @method
+   * @function
```
what's with these changes? not a big deal I think but they _are_ methods; I don't think jsdoc4readme is going to hiccup on these but are they necessary changes?

achingbrain replied:
That's the linter that's done that, otherwise you get:

```
Invalid JSDoc tag (preference). Replace "method" JSDoc tag with "function". [Error/jsdoc/check-tag-names]
```

Apparently there's [no such tag](https://jsdoc.app/#block-tags) as `@method`, but `@function` is marked:

```
[@function](https://jsdoc.app/tags-function.html) (synonyms: @func, @method)
Describe a function or method.
```

rvagg replied:
amusing, since it's right there in the `synonyms`

that's fine, if jsdoc4readme has a problem with it I should be able to address it there

rvagg commented on test/test-indexer.spec.js:
```diff
@@ -78,10 +80,12 @@ describe('CarIndexer fromIterable()', () => {
     return verifyIndexer(indexer)
   })
 
+  // fails on chrome, not sure why
```
even with the changes below? so we need to fix this before we can proceed?

achingbrain replied:
Oops, forgot to remove that comment.

rvagg reviewed (changes requested):
* .gitignore .coverage/ and remove from commit
* there's now 2 TOCs in the README, jsdoc4readme generates its own TOC for the API section; can we tell aegir to limit the TOC depth so we only get a top-level TOC at the top?
* are the lint fixes in the README automatic? jsdoc4readme is pulling that markdown out of the source to insert there, so you've edited markdown that's going to be re-generated when jsdoc4readme is run next—this is OK, if the fixes are automatic and they can be layered on top of it; but if they're manual edits then we'll probably need to reconcile the changes in the source

achingbrain commented:
- I've removed the coverage dir
- Also removed the extra TOC
- The README linting is just how the [remark](https://www.npmjs.com/package/remark) AST to MarkDown likes it - I've reverted pretty much all of the changes, they're not massively important for this PR

rvagg commented on package.json:
```diff
@@ -2,63 +2,112 @@
-    "./decoder": {
-      "import": "./lib/decoder.js"
+    "./buffer-writer": {
+      "types": "./src/buffer-writer.d.ts",
```
I think these all need `./dist` in them, from this one downward; I don't believe that's taken care of by `typesVersions` mapping is it?

achingbrain replied:
Yes - have updated it.

rvagg commented:
@achingbrain the next problem I have with this is I can't run the electron tests when I `npm t` this; most of my dev is on a Linux machine and mostly without a local DISPLAY; so properly headless. Does there need to be an a special xvfb incantation for this; can I get by without having to install anything special? And if it does require special pieces then we'll need to document that for others who might be contributing; _or_ we remove those from the `test` script and just do it in CI (I'd be fine with that, locally I just want basic confirmation and am happy to defer full test confirmation to CI).

achingbrain commented:
@rvagg I guess you'd need to run it via [xvfb-maybe](https://www.npmjs.com/package/xvfb-maybe) for your setup?  Sounds fiddly to me.

We could just make the test command do the default `aegir test` - that would test in node, browsers and webworkers which would be consistent with other libp2p/ipfs modules.  The other `test:*` scripts are available for testing other platforms and CI is the gatekeeper for PRs etc.

It does linting as well at the moment which I think is your preference?

rvagg commented:
@achingbrain yeah, although I'm not too fussed if we do comprehensive in CI. Basic Node, maybe at least one (easy) browser, linting and typing would be my preference but really I just want people to (a) not have too many hurdles to getting a contribution up but (b) lower the chance of hitting a CI problem with their contribution. So `aegir test` is the same as what's there now but without `-t electron-main`?

achingbrain commented:
> So aegir test is the same as what's there now but without -t electron-main?

Yes, exactly - I've simplified it to that so it'll just lint, run tests in node/browsers/webworkers and test the examples.
